### PR TITLE
Download cmdline-interface improvement

### PR DESCRIPTION
**Pull request by oed**
This should address issue #449.

### CODE REVIEW

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -911,12 +911,22 @@ def toggle_inspector(self):
                 cur.inspector.show()
 
     @cmdutils.register(instance='command-dispatcher', scope='window')
-    def download_page(self):
-        """Download the current page."""
-        page = self._current_widget().page()
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Good idea! But please keep `:download-page` for now, and add a `deprecated="Use :download instead."` to the `@cmdutils.register` decorator. This will keep it hidden everywhere, but it still works for now.

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -911,12 +911,22 @@ def toggle_inspector(self):
...
+        """Download a given URL, or current page if none given.
+
+        Args:
+            url: The URL to download, or None to download current page.
+            dest: The file path to write the download to, or None to ask.
+        """
         download_manager = objreg.get('download-manager', scope='window',
                                       window=self._win_id)
-        download_manager.get(self._current_url(), page)
+        if (url):
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: No parens here please (`if url:`).

File: qutebrowser/browser/downloads.py [hunk 3]
```diff
@@ -644,7 +632,7 @@ def get_request(self, request, page=None, fileobj=None, filename=None,
         request.setAttribute(QNetworkRequest.CacheLoadControlAttribute,
                              QNetworkRequest.AlwaysNetwork)
         if fileobj is not None or filename is not None:
-            return self.fetch_request(request, filename, fileobj, page,
+            return self.fetch_request(request, page, fileobj, filename,
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: Good catch! :)

File: qutebrowser/browser/commands.py [hunk 4]
```diff
@@ -911,12 +911,22 @@ def toggle_inspector(self):
...
     @cmdutils.register(instance='command-dispatcher', scope='window')
-    def download_page(self):
-        """Download the current page."""
-        page = self._current_widget().page()
+    def download(self, url=None, dest=None):
+        """Download a given URL, or current page if none given.
+
+        Args:
+            url: The URL to download, or None to download current page.
+            dest: The file path to write the download to, or None to ask.
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Here `None` makes sense from the developer's perspective, but not if an user reads it in the command documentation. Probably *not given* or so would make more sense?

File: qutebrowser/browser/downloads.py [hunk 5]
```diff
@@ -820,6 +819,8 @@ def can_clear(self):
         else:
             return False
 
+    @cmdutils.register(instance='download-manager', name='downloads-clear',
+                        scope='window')
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]:
```
qutebrowser/browser/downloads.py:823:25: E127 continuation line over-indented for visual indent
```

;)

**Comment by The-Compiler** [comment]
I think it'd make much more sense for `:download-open` to open the last finished download rather than the last in the list, what do you think? That'd be a bit more complicated though, as it has to account for finished but already removed downloads in the list.

Also maybe `:downloads-clear` should be renamed to `:download-remove` or so, also remove the last finished by default, and take `--all` as flag or so?

File: qutebrowser/browser/downloads.py [hunk 6]
```diff
@@ -745,20 +733,31 @@ def fetch(self, reply, fileobj=None, filename=None, auto_remove=False):
         return download
 
     @cmdutils.register(instance='download-manager', scope='window')
-    def cancel_download(self, count: {'special': 'count'}=1):
-        """Cancel the first/[count]th download.
+    def download_cancel(self, count: {'special': 'count'}=0):
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: Same here - please keep `cancel_download` around for now, with `deprecated="Use :download-cancel instead!"` or so.

**Comment by oed** [comment]
Wouldn't opening a file that not is in the list be confusing? Or am I missing something?

**Comment by The-Compiler** [comment]
Hm, okay, what to do if people have set `ui -> remove-finished-downloads` is debatable, I guess then your behaviour makes more sense indeed.

But for example if I have three downloads, and the middle one finishes, I'd expect `:download-open` to open the middle one, not to attempt to open the right one.

**Comment by oed** [comment]
Yeah, I agree with that. Will look in to that later today.

**Comment by oed** [comment]
Do you think it makes sense to just have the :download command with the arguments:
start - the same as :download was before
open - open a download
cancel - cancel a download
remove - remove a download
Then the --all flag could be used for all but the start argument.

**Comment by The-Compiler** [comment]
I thought about such sub-commands in general, but I think that's something for later - first `src2asciidoc` (the thing generating the documentation) and how the completion works would have to be adjusted. So I think for now let's keep it in separate commands.

**Comment by The-Compiler** [comment]
I meant something a bit different with "the last finished download" - let me try to explain again:

Let's say there are two downloads running:

```
[foo 70%] [bar 90%]
```

Now the right one finishes (e.g. while the user is not watching):

```
[foo 70%] [bar 100%]
```

Now (e.g. while the user is looking at the window), the left one finishes:

```
[foo 100%] [bar 100%]
```

Now which one would you expect `:download-open` to open? I'd say the left one, i.e. the *most recently finished* one, as opposed to the rightmost (most recently started) one..

But I'm not sure if the benefit is big enough to implement this (keeping a list of most recently finished downloads, ensuring they're still garbage-collected properly, etc.).

What do you think?

Also I've only looked at the diff so far, which looks good - I hope I'll have time to test and merge this tomorrow or so.

**Comment by oed** [comment]
I think it makes more sense to open the download that you started most recently, i.e. the one to the right.

**Comment by oed** [comment]
Irc log regarding this issue (fixed some typos):
```
22:41 < The-Compiler> oed: what do you think tough - maybe it'd be better to 
                      number all downloads for :download-open, and then also show 
                      a download's index in the view like dwb does, so the user 
                      doesn't have to count?
22:43 < oed> yeah, that might be a good idea
22:43 < oed> but then it would make no sense for  :download-open to only open 
             finished downloads
22:44 < oed> maybe :download-open could just flash an error saying that the 
             download isn't done?
22:52 < The-Compiler> oed: +1
22:54 < oed> and something similar in :download-remove I guess
22:58 < The-Compiler> oed: Not sure about :download-remove - shouldn't it just 
                      cancel running downloads, and remove finished ones?
22:59 < oed> then there would be no reason to have :download-cancel?
23:08 < The-Compiler> oed: hm, indeed. Okay, maybe keep them separate, so users 
                      don't accidentally cancel downloads by mistyping a number
23:09 < oed> it should maybe be more clear that :download-remove only removes a 
             download from the ui also
```

**Comment by isene** [comment]
If my vote counts toward anything, I would vote for opening the first
download Started - because that is something the user would know (as
opposed to the first or last download Finished which may be beyond the
user's knowledge [is he is away from the screen while several downloads
finish])


On Wed, Feb 11, 2015 at 01:22:46PM -0800, Joel Torstensson scribed:
> I think it makes more sense to open the download that you started most
> recently, i.e. the one to the right.
> 
> —
> Reply to this email directly or view it on GitHub.*
> 

-- 
		       "Tror du på virkeligheten?"
				   (Jonatan, 5 år)

Geir Isene                 |    /"\    ASCII Ribbon Campaign     
http://isene.com           |    \ /    Respect for open standards
http://isene.me            |     X     No HTML/RTF in email      
+47-4550 2470              |    / \    No MS Word docs in email

File: qutebrowser/browser/downloads.py [hunk 7]
```diff
@@ -873,6 +926,12 @@ def remove_items(self, downloads):
                 download.deleteLater()
         self.endRemoveRows()
 
+    def update_indexes(self):
+        """Update indexes of all DownloadItems"""
+        for i, d in enumerate(self.downloads, 1):
+            if not d.index == i:
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: This should be written `if d.index != i:` instead.

File: qutebrowser/browser/downloads.py [hunk 8]
```diff
@@ -873,6 +926,12 @@ def remove_items(self, downloads):
...
+                d.index = i
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: Actually, ignore the previous comment - I'd just update the index unconditionally.

File: qutebrowser/browser/downloads.py (hunk 7 above)
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: I'm not sure if this is still needed (see comment below), but if so, it should be written as `if d.index != i:` instead.

File: qutebrowser/browser/downloads.py (hunk 8 above)
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: You could just alter the indices unconditionally as it doesn't really matter.

However, I believe you have to emit the [dataChanged](http://doc.qt.io/qt-5/qabstractitemmodel.html#dataChanged) signal (see `on_data_changed()` for an example) so it's ensured the view is actually updated. That signal takes two [QModelIndex](http://doc.qt.io/qt-5/qmodelindex.html)es, so you'll need to find out what range of items was updated.

File: qutebrowser/browser/downloads.py [hunk 9]
```diff
@@ -745,20 +733,42 @@ def fetch(self, reply, fileobj=None, filename=None, auto_remove=False):
...
+        """Open the last/[count]th finished download.
+
+        Args:
+            count: The index of the download to cancel.
+        """
+        finished_items = [d for d in self.downloads if d.done and d.successful]
+        try:
+            download = finished_items[count - 1]
+        except IndexError:
+            raise cmdexc.CommandError("There's no download {}!".format(count))
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: All those error messages printing `count` could probably be improved when count is `0`, probably they should only print *There's no dowload!* in that case (*There's no download 0!* is rather confusing).

File: qutebrowser/browser/downloads.py [hunk 10]
```diff
@@ -744,20 +741,78 @@ def fetch(self, reply, fileobj=None, filename=None, auto_remove=False):
...
 
+    def raise_no_download(self, index):
+        """Raise an exception that the download doesn't exist
+
+        Args:
+            index: The index of the download
+        """
+        if not index:
+            raise cmdexc.CommandError("There's no download!")
+        raise cmdexc.CommandError("There's no download {}!".format(index))
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: I suggest renaming `index` to `count` to avoid confusion (count starts with 1, indices start with 0).

File: qutebrowser/browser/downloads.py [hunk 11]
```diff
@@ -873,6 +946,18 @@ def remove_items(self, downloads):
...
         self.endRemoveRows()
 
+    def update_indexes(self):
+        """Update indexes of all DownloadItems"""
+        first_idx = None
+        for i, d in enumerate(self.downloads, 1):
+            if first_idx is not None and d.index != i:
+                first_idx = i - 1
+            d.index = i
+        if first_idx:
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: I think this should be `if first_idx is not None:` as well.

File: qutebrowser/browser/downloads.py [hunk 12]
```diff
@@ -873,6 +946,18 @@ def remove_items(self, downloads):
                 download.deleteLater()
         self.endRemoveRows()
 
+    def update_indexes(self):
+        """Update indexes of all DownloadItems"""
+        first_idx = None
+        for i, d in enumerate(self.downloads, 1):
+            if first_idx is not None and d.index != i:
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: If I'm not mistaken, this will never set `first_idx` as it will always be `None`.

**Comment by The-Compiler** [comment]
I just gave it a quick try - works well and it feels very intuitive to use! Please address the latest 3 comments, then I'll take a closer look again and we can merge this! :)

File: qutebrowser/browser/downloads.py (hunk 11 above)
Comment (oed) [inline qutebrowser/browser/downloads.py]: I'm not exactly sure what you mean by that.

File: qutebrowser/browser/downloads.py (hunk 12 above)
Comment (oed) [inline qutebrowser/browser/downloads.py]: Yes, it should be if first_idx is None and d.index != i:

File: qutebrowser/browser/downloads.py (hunk 11 above)
Comment (oed) [inline qutebrowser/browser/downloads.py]: Oh, got it!
